walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,132 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
+
+  // Returns available times for meeting requesters based on the events and its attendees' schedules.
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> optionalAttendees = new HashSet(request.getOptionalAttendees());
+    Set<String> allAttendees = getAllAttendees(events);
+    List<TimeRange> availableTimeRanges = new ArrayList();
+
+    List<Event> eventsListOfMandatory = createSortedEventsListOfMandatory(events, optionalAttendees);
+    List<Event> eventsListOfMandatoryAndOptional = createSortedEventsListOfMandatoryAndOptional(events);
+
+    if (isRequestDurationTooLong(request)) {
+      return availableTimeRanges;
+    }
+
+    if (hasNoOverlappingAttendees(request.getAttendees(), allAttendees)) {
+      availableTimeRanges.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+      return availableTimeRanges;
+    }
+
+    List<TimeRange> timeRangesList = findTimeRangesOfEvents(eventsListOfMandatoryAndOptional);
+    removeNestedAndOverlappedEvents(timeRangesList);
+    addAvailableTimeRanges(availableTimeRanges, timeRangesList, request);
+
+    if (availableTimeRanges.size() == 0) {
+      List<TimeRange> timeRangesListOnlyMandatory = findTimeRangesOfEvents(eventsListOfMandatory);
+      removeNestedAndOverlappedEvents(timeRangesListOnlyMandatory);
+      addAvailableTimeRanges(availableTimeRanges, timeRangesListOnlyMandatory, request);
+    }
+    return availableTimeRanges;
+  }
+
+  // Returns a set of all of the attendees of events.
+  private Set<String> getAllAttendees(Collection<Event> events) {
+    Set<String> allAttendees = new HashSet();
+    for (Event event: events) {
+      allAttendees.addAll(event.getAttendees());
+    }
+    return allAttendees;
+  }
+
+  // Checks if the requesters of a meeting and the attendees have overlapping members.
+  private boolean hasNoOverlappingAttendees(Collection<String> requesters, Set<String> attendees) {
+    return Collections.disjoint(requesters, attendees);
+  }
+
+  // Sorts the TimeRanges of events of only mandatory attendees by their start time.
+  private List<Event> createSortedEventsListOfMandatory(Collection<Event> events, Set<String> optionalAttendees) {
+    List<Event> eventsList = new ArrayList(events);
+    eventsList.removeIf(event -> event.getAttendees().equals(optionalAttendees));
```

Comment:
Lots easier to read! Do you want to remove on if the event === the optional attendees?

I think you want to remove if optionalAttendees contains the full set of event.getAttendees()

Let's say A B and C are optional. 

if an event is just B, then it should be removed too right?

Reply:
Also, this can contain more than just the mandatory right? Wouldn't the correct check be:

// pass in mandatory instead of optional ? ie,

remove if ( !event.getAttendees().intersects(mandatoryAttendees) )

Reply:
This makes perfect sense! Thank you for this suggestion! I wasn't able to figure out the exact syntax of the parameter for removeIf, because removeIf( event -> !event.getAttendees().intersects(mandatoryAttendees)) did not work. I instead used Collections.disjoint().